scripts/fetch_data.py

```diff
@@ -0,0 +1,274 @@
+        'djf': ['0', 'seasonal'],
+        'mam': ['1', 'seasonal'],
+        'jja': ['2', 'seasonal'],
+        'son': ['3', 'seasonal'],
+        'jan': ['0', 'monthly'],
+        'feb': ['1', 'monthly'],
+        'mar': ['2', 'monthly'],
+        'apr': ['3', 'monthly'],
+        'may': ['4', 'monthly'],
+        'jun': ['5', 'monthly'],
+        'jul': ['6', 'monthly'],
+        'aug': ['7', 'monthly'],
+        'sep': ['8', 'monthly'],
+        'oct': ['9', 'monthly'],
+        'nov': ['10', 'monthly'],
+        'dec': ['11', 'monthly'],
+    }[time_of_year]
+    args['time'] = ce_time
+    args['timescale'] = ce_timescale
+
+    ce_cell_method = {
+        'iastddev': 'standard_deviation',
+        'iamean': 'mean'
+    }[temporal]
+    args['cell_method'] = ce_cell_method
+
+    ce_spatial = {
+        's0p': 'min',
+        's100p': 'max',
+        'smean': 'mean'
+    }[spatial]
+    args['spatial'] = ce_spatial
+
+    ce_percentile = {
+        'e25p': 25,
+        'e75p': 75,
+        'hist': 100
+    }[percentile]
+    args['percentile'] = ce_percentile
+
+    if percentile == 'hist':
+        emission = percentile
+    else:
+        emission = variable
+    ce_emission = {
+        'temp': 'historical,rcp85',
+        'prec': 'historical,rcp85',
+        'dg05': 'historical,rcp85',
+        'nffd': 'historical, rcp85',
+        'pass': 'historical,rcp85',
+        'dl18': 'historical,rcp85',
+        'hist': ''  # historical has no emission scenario
+    }[emission]
+    args['emission'] = ce_emission
+
+    args['area'] = area['the_geom']
+
+    if percentile == 'hist':
+        dates = percentile
+    else:
+        dates = date_range
+    ce_dates = {
+        'hist': ['19710101-20001231'],
+        '2020': ['20100101-20391231',
+                 '20110101-20400101',
+                 '20100101-20391230'],
+        '2050': ['20400101-20691231',
+                 '20410101-20700101',
+                 '20400101-20691230'],
+        '2080': ['20700101-20991231',
+                 '20710101-21000101',
+                 '20700101-20991230']
+    }[dates]
+    args['dates'] = ce_dates
+
+    return args
+
+
+def get_ce_data(sesh, var_name, date_range, area):
+    """Parse a given variable name and into parameters that are used to query
+       the Climate Explorer backend.
+    """
+    try:
+        variable, time_of_year, temporal, \
+            spatial, percentile = var_name.split('_')
+    except ValueError as e:
+        logger.error('Error: Unable to read variable name {} error: {}'
+                     .format(var_name, e))
+        return None
+
+    args = prep_args(variable, time_of_year, temporal, spatial,
+                     percentile, area, date_range)
+    models = get_models(sesh, percentile)
+
+    result = [
+        data for data in [query_backend(sesh, model, args) for model in models]
+        if data is not None
+    ]
+
+    return np.asscalar(np.percentile(result, args['percentile']))
```

Comment:
I'm commonly getting errors on this line with a traceback such as this:

```
Traceback (most recent call last):
  File "scripts/process.py", line 115, in <module>
    rules = resolve_rules(args.csv, args.date_range, region, args.log_level)
  File "/local_temp/hiebert/p2a-rule-engine/scripts/resolver.py", line 50, in resolve_rules
    for var in variables}
  File "/local_temp/hiebert/p2a-rule-engine/scripts/resolver.py", line 50, in <dictcomp>
    for var in variables}
  File "/local_temp/hiebert/p2a-rule-engine/scripts/fetch_data.py", line 275, in get_variables
    return get_ce_data(sesh, var_name, date_range, area)
  File "/local_temp/hiebert/p2a-rule-engine/scripts/fetch_data.py", line 267, in get_ce_data
    return np.asscalar(np.percentile(result, args['percentile']))
  File "/local_temp/hiebert/env/lib/python3.6/site-packages/numpy/lib/function_base.py", line 3702, in percentile
    a, q, axis, out, overwrite_input, interpolation, keepdims)
  File "/local_temp/hiebert/env/lib/python3.6/site-packages/numpy/lib/function_base.py", line 3821, in _quantile_unchecked
    interpolation=interpolation)
  File "/local_temp/hiebert/env/lib/python3.6/site-packages/numpy/lib/function_base.py", line 3400, in _ureduce
    r = func(a, **kwargs)
  File "/local_temp/hiebert/env/lib/python3.6/site-packages/numpy/lib/function_base.py", line 3936, in _quantile_ureduce_func
    x1 = take(ap, indices_below, axis=axis) * weights_below
  File "/local_temp/hiebert/env/lib/python3.6/site-packages/numpy/core/fromnumeric.py", line 189, in take
    return _wrapfunc(a, 'take', indices, axis=axis, out=out, mode=mode)
  File "/local_temp/hiebert/env/lib/python3.6/site-packages/numpy/core/fromnumeric.py", line 56, in _wrapfunc
    return getattr(obj, method)(*args, **kwds)
IndexError: cannot do a non-empty take from an empty axes.
```

This could probably use some error checking and crash hardening around it. I'm guessing that result is coming back with bad data...

Reply:
Ya I think there needs to be better error handling here.  The error is caused by `result` being None/Empty.  I think in general the handling of rules that cannot be resolved could be improved.